src/client/driver/driver-link/child.js

```diff
@@ -0,0 +1,98 @@
+import { Promise, eventSandbox, nativeMethods } from '../deps/hammerhead';
+import { domUtils, delay } from '../deps/testcafe-core';
+import { CurrentIframeIsNotLoadedError, CurrentIframeNotFoundError, CurrentIframeIsInvisibleError } from '../../../errors/test-run';
+import { ensureElementVisible } from '../ensure-element-utils';
+import sendMessageWithConfirmation from './send-message-with-confirmation';
+import INTER_DRIVER_MESSAGES from './messages';
+import DriverStatus from '../status';
+
+
+const CHECK_IFRAME_EXISTENCE_INTERVAL = 1000;
+const WAIT_IFRAME_RESPONSE_DELAY      = 500;
+
+
+export default class ChildDriverLink {
+    constructor (driverWindow, driverId, iframeAvailabilityTimeout) {
+        this.driverWindow              = driverWindow;
+        this.driverIframe              = domUtils.findIframeByWindow(driverWindow);
+        this.driverId                  = driverId;
+        this.iframeAvailabilityTimeout = iframeAvailabilityTimeout;
+    }
+
+    _ensureIframe () {
+        if (!domUtils.isElementInDocument(this.driverIframe))
+            return Promise.reject(new CurrentIframeNotFoundError());
+
+        return ensureElementVisible(this.driverIframe, this.iframeAvailabilityTimeout, () => new CurrentIframeIsInvisibleError());
+    }
+
+    _waitForIframeRemovedOrHidden () {
+        return new Promise(resolve => {
+            this.checkIframeInterval = nativeMethods.setInterval.call(window,
+                () => this._ensureIframe().catch(resolve), CHECK_IFRAME_EXISTENCE_INTERVAL);
+        });
+    }
+
+    _waitForCommandResult () {
+        return new Promise(resolve => {
+            var resolved = false;
+
+            var onMessage = e => {
+                var msg = e.message;
+
+                /*eslint-disable no-use-before-define*/
+                if (msg.cmd === INTER_DRIVER_MESSAGES.onCommandExecuted)
+                    tearDownAndResolve(msg.status);
+                /*eslint-enable no-use-before-define*/
+            };
+
+            var tearDownAndResolve = status => {
+                if (resolved)
+                    return;
+
+                resolved = true;
```

Comment:
You can turn `onMessage` into Promise and then use `Promise.race` for message and other part of functionality. This will help to get rid of `tearDownAndResolve` arrow and awkward `resolved` var